README.md
poi
=====

yay a 2d poi pattern grapher!

hella work in progress

practical todos
-----
+ ~~selectable patterns~~
+ ~~split/together/same time selector~~ (pattern phase shift)
+ ~~rotate the graph to make theta=0 point up~~
+ more patterns. [ideas](https://github.com/infiniteperplexity/visual-spinner-3d/tree/master/json)
+ ~~refactor so the speed is changeable~~
+ ~~adjustable poi to arm ratio~~
+ ~~urlparams for permalinking~~

pipe dreams / ideas
-----
+ FORCE DIAGRAMS AND GRAVITY OHHH SNAP
+ beat detection and counter
+ ~~RAVE MODE~~
+ Fourier transform arbitrary shapes into what you have to do with your hands over a long exposure and tell you how many cycles it would take
+ slight hand offsets and positioning relative to body? I want this to make it easier for noobs like me to figure out how their hands should interact with each other especially when crossing over
+ write a constraint solver to animate someone's arms attached to the handles? this would be fun and silly but might discourage creativity and good body movement
+ ~~allow arbitrary pattern phase shifts and n-flowers (I'm writing the pattern generators pretty modularly so this should follow naturally)~~
+ refactor to make specifications for stalls and other directional changes easy, probably some kind of chaining mechanism that auto-normalizes itself to overall period.
+ implement "your arms can't do that" detector (involves reverse engineering sneaky ways in which people fold poi around their bodies)
+ idk what if I made a polar equation validator that people could input patterns into so I don't have to figure all of them out myself


CHANGELOG.md
2016-07-22
-----
+ adjustable poi length

2016-07-21
-----
+ permalinking

2016-07-20
-----
+ allow full-pattern rotations
+ allow pattern phase shifts

2016-07-18
-----
+ make parametric pattern generators accessible in interface

2016-07-13
-----
+ configuration options for display
+ swappable patterns
+ RAVE MODE

2016-07-12
-----
initial commit. wrote a polar plotter and some basic patterns


app.js
var app = angular.module('PoiApp', []);


function repeat(o, n) {
  /* equivalent to pythonic [o]**n */
  return _.map(_.range(n), function(i) { return o; });
}


function gcd(a, b) {
  if (a == 0) { return b; }
  while (b != 0) {
    if (a > b) {
      a = a - b;
    } else {
      b = b - a;
    }
  }
  return a;
}


app.controller('PoiCtrl', function($scope, $http) {

  $scope.NULL_SELECT_VALUE = '---';
  $scope.options = {};

  // this construction assumes that pattern names and pattern generator
  // names never overlap; TODO should enforce this or add a delineator
  $scope.patternNameOptions = Array.prototype.concat(
    [$scope.NULL_SELECT_VALUE],
    _.keys(patterns),
    _.keys(pattern_generators)
  );

  /* Pattern initialization */

  // Length should not change, please pretend this is statically allocated -_-
  $scope.selectedPatternNames = [null, null];

  // Generator parameters, if selected
  $scope.patternGeneratorOptions = repeat({}, $scope.selectedPatternNames.length);

  var DEFAULT_PATTERNS = ['n_petal_antispin', 'n_petal_inspin'];

  /* Convenience functions & checkers */

  $scope.getPatterns = function(patternNames) {
    if (!patternNames) { patternNames = $scope.selectedPatternNames; }
    return _.map(
      _.filter(patternNames, function(p) {
        return p in patterns;
      }),
      function(p) { return patterns[p]; });
  };

  $scope.getPatternColor = function(patternIndex) {
    return settings.point_colors[patternIndex % settings.point_colors.length];
  };

  $scope.isSelectedPatternGenerator = function(i) {
    /* Is the i'th pattern a generator */
    return $scope.selectedPatternNames[i] in pattern_generators;
  }

  // Given a param specification, return an array of its value options
  $scope.getParamValues = function(argSpec) {
    if (_.isArray(argSpec.values)) { return argSpec.values; }
    _.each(['step', 'start', 'stop'], function(prop) {
      argSpec[prop] = parseInt(argSpec[prop]);
      if (!_.isNumber(argSpec[prop]) || _.isNaN(argSpec[prop])) {
        argSpec[prop] = 1;
      }
    });
    return _.range(argSpec.start, argSpec.stop + argSpec.step, argSpec.step);
  };

  /* Stats */

  $scope.getBeatRatio = function() {
    /* return human-readable string representing beat ratio in lowest terms
     * TODO: figure out what to do for >2 poi?
     * TODO: this is pretty sloppy -- go blackbox it with where the patterns
     *  are stored plz
     */
    if (_.some($scope.renderer.patterns, function(p) { return !p; })) {
      return null;
    }
    var beat_gcd = gcd($scope.renderer.patterns[0].beats,
                  $scope.renderer.patterns[1].beats);
    return _.map(_.pluck($scope.renderer.patterns, 'beats'), function(x) {
      return x / beat_gcd;
    }).join(':');
  };

  /* Methods that pass configuration changes to the renderer */

  $scope.handlePatternSelect = function(i, preventUrlUpdate) {
    /* Handler for pattern selector */
    var pattern_name = $scope.selectedPatternNames[i];
    if (!pattern_name || pattern_name == $scope.NULL_SELECT_VALUE) {
      $scope.patternGeneratorOptions[i] = {};
      $scope.updatePattern(i);
      if (!pattern_name) { return; }
    }
    // again, this assumes pattern names and generator names don't overlap...
    // Selected a generative pattern; set up its interface options
    else if (pattern_name in pattern_generators) {
      var generator = pattern_generators[pattern_name];
      $scope.patternGeneratorOptions[i] = {
        name: pattern_name,
        args: _.pluck(generator.args, 'default'),
        argNames: _.pluck(generator.args, 'name'),
        argSpecs: generator.args,
      };
      $scope.showGeneratedPattern(i);
    }
    // Selected pattern preset; directly update renderer
    else if (pattern_name in patterns) {
      $scope.patternGeneratorOptions[i] = {};
      $scope.updatePattern(i, patterns[pattern_name]);
    }
    if (!preventUrlUpdate) { $scope.saveStateToUrlParams(); }
  };

  $scope.showGeneratedPattern = function(i, preventUrlUpdate) {
    /* Handler to update the renderer once a generator's params are set */
    var generator = pattern_generators[$scope.selectedPatternNames[i]];
    if (!generator) { return; }
    var pattern = generator.generator.apply(
      null, $scope.patternGeneratorOptions[i].args);
    $scope.renderer.patterns[i] = pattern;
    if (!preventUrlUpdate) { $scope.saveStateToUrlParams(); }
  };

  $scope.getPatternParamCallback = function(patternIndex, paramIndex) {
    /* Return a callback which updates the value for param `paramIndex` for
     * the `patternIndex`th generated pattern.
     */
    return function(value) {
      $scope.patternGeneratorOptions[patternIndex].args[paramIndex] = value;
      $scope.showGeneratedPattern(patternIndex);
    };
  };

  $scope.updatePattern = function(i, pattern) {
    /* Updates the i'th pattern in the renderer */
    // TODO: force renderer to have fixed size patterns array
    pattern = pattern || patterns[i];
    $scope.renderer.patterns[i] = pattern;
  };

  $scope.setSpeed = function() {
    var multiplier = Math.pow(1.5, -$scope.options.speed);
    options.REFRESH = 25 * multiplier;
    clearInterval($scope.loop);
    $scope.runMainLoop();
  };

  $scope.pauseHandler = function() {
    if ($scope.paused) {
      clearInterval($scope.loop);
    } else {
      $scope.setSpeed();
    }
  };

  $scope.setOption = function(option, value, preventUrlUpdate) {
    // These options are dynamically read by the plotter, so we don't need
    // to do anything else. I know, it's gross :x
    window.options[option] = (value===undefined) ? $scope.options[option] : value;
    if (!preventUrlUpdate) { $scope.saveStateToUrlParams(); }
  };

  $scope.setOptions = function(options, preventUrlUpdate) {
    if (!preventUrlUpdate) { $scope.saveStateToUrlParams(); }
    $scope.options = options;
    _.each($scope.options, function(val, key) {
      $scope.setOption(key, val, preventUrlUpdate);
    });
  };

  /* Permalinking */

  $scope.getUrlParams = function() {
    var params = _.object(_.map(
      window.location.search.substring(1).replace(/\/$/, '').split('&'),
      function(pair) { return pair.split('='); }
    ));
    return params;
  };

  // TODO: this is attached to a lot of updaters with a preventUpdate arg.
  // Would be nicer to figure out how to write a JS decorator that can require
  // an additional arg
  $scope.saveStateToUrlParams = function() {
    var params = {
      options: JSON.stringify($scope.options),
      patterns: JSON.stringify($scope.selectedPatternNames),
      args: JSON.stringify(_.pluck($scope.patternGeneratorOptions, 'args')),
    };
    window.location.hash = $.param(params);
  };

  $scope.loadStateFromUrlParams = function() {
    var params = _.object(_.map(
      $.deparam(window.location.hash.substring(1)),
      function(val, key) { return [key, JSON.parse(val)]; }
    ));
    $scope.setOptions(params.options || {}, true);
    $scope.selectedPatternNames = params.patterns || DEFAULT_PATTERNS;
    params.args = (params.args ||
                   repeat(null, $scope.selectedPatternNames.length));
    _.each($scope.selectedPatternNames, function(name, i) {
      $scope.handlePatternSelect(i);   // setup pattern
      if (params.args[i]) {
        $scope.patternGeneratorOptions[i].args = params.args[i];  // override args
        $scope.showGeneratedPattern(i);  // update renderer
      }
    });
  };

  /* Main */

  $scope.runMainLoop = function() {
    $scope.loop = setInterval(function() {
      $scope.renderer.draw($scope.theta, $scope.r);
      $scope.advanceTime();
    }, options.REFRESH);
  };

  $scope.initialize = function() {
    var canvas = $('#canvas').get(0),
        ctx = canvas.getContext("2d"),
        origin = new Vector(300, 300),
        initial_patterns = $scope.getPatterns();
    $scope.r = 150;
    $scope.theta = 0;

    $scope.renderer = new TravelingPlotter(
      ctx, origin.x, origin.y, initial_patterns);

    settings.canvas.width = canvas.width;
    settings.canvas.height = canvas.height;

    $scope.runMainLoop();

    $scope.advanceTime = function() {
      $scope.theta += get_d_theta();
      if ($scope.theta > 2*Math.PI) { $scope.theta = 0; }
    };

    _.each($scope.selectedPatternNames, function(name, i) {
      $scope.handlePatternSelect(i);
    });

    // tmp test
    $scope.loadStateFromUrlParams();
  }; $scope.initialize();
});


/* Directive for config checkbox that passes option to renderer onchange
 * :attr label: label to display
 * :attr name: config variable name
 * :attr value: is checked?
 */
app.directive('controlCheckbox', function() {
  return {
    restrict: 'A',
    template: function(element, attrs) {
      return '' +
      '<span class="control">' + attrs.label +
      '  <input type="checkbox" ' +
      '         ng-model="options.' + attrs.name + '" ' +
      '         ng-init="options.' + attrs.name + '= ' + attrs.value + '; ' +
      '                  setOption(\'' + attrs.name + '\')" ' +
      '         ng-change="setOption(\'' + attrs.name + '\')" ' +
      '         />' +
      '</span>';
    },
  };
});


index.js
// Plotter loosely based on https://gist.github.com/kazad/8bb682da198db597558c,
// totally different implementation though

var settings = {
  REFRESH: 25,   // refresh rate in ms
  STEPS: 180,    // # of intervals to divide wave into

  canvas: {
    width: 0,
    height: 0,
  },

  origin: {
    color: '#ffa500',
    size: 5,
    stroke_color: '#c2a7dd',
  },

  point: {
    color: '#f00',
    size: 10,
    stroke_color: '#fcc',
  },

  point_colors: ['#f00', '#00f', '#0f0'],  // lol hopefully we don't have more than 3 poi
};

settings.d_theta = 2 * Math.PI / settings.STEPS;

window.options = {};


// Hooray, pseudo-OOP JS

/* Canvas graphics primitives */

var CanvasRenderer = function(ctx) {
  this.ctx = ctx;
};
_.extend(CanvasRenderer.prototype, {

  reset: function() {
    this.ctx.clearRect(0, 0, settings.canvas.width, settings.canvas.height);
  },

  draw_circ: function(x, y, r) {
    this.ctx.beginPath();
    this.ctx.arc(x, y, r, 0, Math.PI * 2, true);
    this.ctx.stroke();
    this.ctx.closePath();
  },

  draw_dot: function(x, y, r) {
    r = r || 3.5;
    this.ctx.beginPath();
    this.ctx.arc(x, y, r, 0, Math.PI * 2, true);
    this.ctx.fill();
    this.ctx.closePath();
  },

  draw_line: function(x0, y0, x1, y1) {
    this.ctx.beginPath();
    this.ctx.moveTo(x0, y0);
    this.ctx.lineTo(x1, y1);
    this.ctx.stroke();
    this.ctx.closePath();
  },

});


/* Plotter with relative origin */

var Plotter = function(ctx, x, y) {
  this.set_origin(x, y);
  Plotter.prototype.__proto__.constructor.apply(this, [ctx]);
};
Plotter.prototype = Object.create(CanvasRenderer.prototype);
_.extend(Plotter.prototype, {
  constructor: Plotter,

  draw: function(theta, r) {
    this.draw_polar_point(theta, r);
  },

  refresh: function(theta, r) {
    this.reset();
    this.draw(theta, r);
  },

  set_origin: function(x, y) {
    this.origin = {x: x, y: y};
  },

  set_point_color: function(color) {
    this.point_color = color;
  },

  /* Convenience drawing methods */

  draw_polar_point: function(theta, radius) {
    this.draw_point(radius * Math.cos(theta), radius * Math.sin(theta));
  },

  draw_point: function(x, y) {
    this.ctx.strokeStyle = '#ccc';
    this.ctx.lineWidth = 1.5;
    this.draw_line_to_origin(x, y);
    
    this.ctx.fillStyle = this.point_color || settings.point.color;
    this.draw_dot(x, y, settings.point.size);
  },

  draw_line_to_origin: function(x, y) {
    this.draw_line(0, 0, x, y);
  },

  /* Overridden primitives */

  draw_circ: function(x, y, r) {
    Plotter.prototype.__proto__.draw_circ.apply(this, [
      x + this.origin.x, y + this.origin.y, r,
    ]);
  },

  draw_dot: function(x, y, r) {
    Plotter.prototype.__proto__.draw_dot.apply(this, [
      x + this.origin.x, y + this.origin.y, r,
    ]);
  },

  draw_line: function(x0, y0, x1, y1) {
    Plotter.prototype.__proto__.draw_line.apply(this, [
      x0 + this.origin.x, y0 + this.origin.y,
      x1 + this.origin.x, y1 + this.origin.y,
    ]);
  },
  
});


/* Polar plotter whose origin travels according to a periodic function.
 *  Traveling function returns a 2D Vector which offsets from
 *  initial origin.
 */

var TravelingPlotter = function(ctx, x, y, patterns) {
  this.patterns = patterns;
  this.initial_origin = {x: x, y: y};
  TravelingPlotter.prototype.__proto__.constructor.apply(this, [ctx, x, y]);
};
TravelingPlotter.prototype = Object.create(Plotter.prototype);
_.extend(TravelingPlotter.prototype, {
  constructor: TravelingPlotter,

  draw: function(theta, r) {
    this.reset();

    var self = this,
        args = arguments;
    _.each(this.patterns, function(pattern, i) {
      self.set_traveling_origin(pattern.traveling_function, theta, r);
      self.set_point_color(settings.point_colors[i % settings.point_colors.length]);

      if (window.options.show_hand_trace) {
        self.trace_origin(pattern, r);
      }
      if (window.options.show_pattern_trace) {
        self.trace_pattern(pattern, r);
      }

      // Draw
      self.constructor.prototype.__proto__.draw.apply(self, [
        pattern.shift_theta(theta), r,
      ]);

      // todo: maybe make sure origins are drawn on top since they're smaller?
      self.draw_origin();
    });
  },

  draw_origin: function() {
    this.ctx.fillStyle = settings.origin.color;
    this.draw_dot(0, 0, settings.origin.size);
  },

  trace_pattern: function(pattern, r) {
    // Persistent trace of a pattern
    this.ctx.strokeStyle = settings.point.stroke_color;
    var circle_fn = pattern_generators.circle();
    this.trace_function(
      function(theta, r) {
        return pattern.traveling_function(theta, r).add(
               circle_fn(pattern.shift_theta(theta), r));
      }, r);
  },

  trace_origin: function(pattern, r) {
    // Persistent trace of the origin's movement
    this.ctx.strokeStyle = settings.origin.stroke_color;
    this.trace_function(pattern.traveling_function, r);
  },

  trace_function: function(fn, r) {
    // Persistent trace of a function(theta, r)
    this.ctx.beginPath();
    this.ctx.lineWidth = 1.5;
    
    // simulate a full period
    // one extra point for smoothness
    for (var theta = 0; theta <= 2 * Math.PI; theta += settings.d_theta) {
      var coords = fn(theta, r);
      if (theta==0) {
        this.ctx.moveTo(this.initial_origin.x + coords.x,
                        this.initial_origin.y + coords.y);
      } else {
        this.ctx.lineTo(this.initial_origin.x + coords.x,
                        this.initial_origin.y + coords.y);
      }
    }
    this.ctx.stroke();
    this.ctx.closePath();
  },

  set_traveling_origin: function(traveling_function, theta, r) {
    var travel = traveling_function(theta, r),
        origin = [this.initial_origin.x + travel.x,
                  this.initial_origin.y + travel.y];
    this.set_origin.apply(this, origin);
  },

});


/* Traveling function signature:
 * :param theta: angle in radians
 * :param d: poi length (handle center to head center)
 * :returns: 2D Vector
 */
var traveling_functions = {

  // simple back and forth oscillation
  test: function(theta, d) {
    return new Vector(theta < Math.PI ? theta : 2*Math.PI - theta, 0);
  },

  // Circle phase shifted by π
  isolation: function(theta, r) {
    theta += Math.PI;
    return new Vector(r * Math.cos(theta), r * Math.sin(theta));
  },

};


// Parametric generators that return functions fitting the format above.
var pattern_generators = {

  /* :param n: number of sides
   * :param rotation: angle to phase shift, in radians, from pointing up
   * :param is_phase_shifted: flips an antispin to an inspin
   */
  polygon: function(n, is_phase_shifted, rotation) {
    var phase_shift = is_phase_shifted ? Math.PI/n : 0;
    rotation = (rotation || 0);
    if (!is_phase_shifted) rotation += Math.PI/n;

    return function(theta, d) {
      theta += rotation;
      // polar equation for edge: cos(π/n) / cos((theta mod 2π/n) - π/n)
      var r = (d * (
        Math.cos(Math.PI/n) / Math.cos((theta + phase_shift) % (2*Math.PI/n) - Math.PI/n)
      ));
      return new Vector(r * Math.cos(theta), r * Math.sin(theta));
    };
  },
  
  circle: function(frequency) {
    frequency = frequency || 1;
    return function(theta, r) {
      theta *= frequency;
      return new Vector(r * Math.cos(theta), r * Math.sin(theta));
    };
  },

};


// Complete pattern specifications
var patterns = {

  test: {
    frequency: -2,
    traveling_function: traveling_functions.isolation,
  },

  extension: {
    frequency: 1,
    traveling_function: pattern_generators.circle(),
  },

  triquetra: {
    frequency: -2,
    phase_shift: Math.PI/3,
    traveling_function: pattern_generators.polygon(3, false),
  },

  four_petal_antispin: {
    frequency: -3,
    phase_shift: Math.PI/2,
    traveling_function: pattern_generators.polygon(4, false, Math.PI/4),
  },

  four_petal_inspin: {
    frequency: -3,
    traveling_function: pattern_generators.polygon(4, true),
  },

};

// struct for pattern specifications (so we get defaults)
function Pattern(options) {
  var self = this;
  this.traveling_function = options.traveling_function;  // must exist
  this.frequency = options.frequency || 1;
  this.phase_shift = options.phase_shift || 0;

  this.shift_theta = function(theta, pattern) {
    return (theta + self.phase_shift) * self.frequency;
  };
}

// map specs to this struct
patterns = _.chain(patterns).map(function(pattern, name) {
  console.log(name, pattern)
  return [name, new Pattern(pattern)];
}).object().value();



// main
$(function() {
  var canvas = $('#canvas').get(0);
  var ctx = canvas.getContext("2d");
  var theta = 0;
  var origin = new Vector(300, 300);
  var r = 125;

  var renderer = new TravelingPlotter(
    ctx, origin.x, origin.y, [
      patterns.extension,
      patterns.triquetra,
      //patterns.four_petal_inspin,
      patterns.four_petal_antispin,
    ]
  );

  settings.canvas.width = canvas.width;
  settings.canvas.height = canvas.height;

  // main loop
  setInterval(function() {
    renderer.draw(theta, r);
    advanceTime();
  }, settings.REFRESH);

  function advanceTime() {
    theta += settings.d_theta;
    if (theta > 2*Math.PI) { theta = 0; }
  };
});


poi.js
// Plotter loosely based on https://gist.github.com/kazad/8bb682da198db597558c,
// totally different implementation though

var settings = {
  canvas: {
    width: 0,
    height: 0,
  },

  origin: {
    color: '#ffa500',
    size: 6,
    stroke_transparency: 0.5,
  },

  point: {
    color: '#f00',
    size: 12,
    stroke_color: '#fcc',
    stroke_transparency: 0.2,
    trail_arc: Math.PI / 4,
  },

  point_colors: ['#f00', '#5430e7', '#0f0'],  // lol hopefully we don't have more than 3 poi
  POINT_DEBUG_COLOR: '#bf00ff',

  rave_mode: {
    background_color: '#000',
    glow_spread: 3,
    glow_blur: 50,
    stroke_transparency: 0.7,
  },

  UNIVERSAL_ROTATION: Math.PI / 2,  // point up instead of right
};

window.options = {
  STEPS: 180,      // # of intervals to divide wave into
  REFRESH: 25,     // refresh rate in ms
  poi_ratio: 0.6,  // poi to arm ratio
};

function get_d_theta() { return 2 * Math.PI / window.options.STEPS };
function to_radians(degrees) { return degrees * 2 * Math.PI / 360; }
function to_degrees(radians) { return radians * 360 / (2 * Math.PI); }


// Hooray, pseudo-OOP JS

/* Canvas graphics primitives */

var CanvasRenderer = function(ctx) {
  this.ctx = ctx;
};
_.extend(CanvasRenderer.prototype, {

  reset: function() {
    if (options.rave_mode) {
      this.ctx.fillStyle = settings.rave_mode.background_color;
      this.ctx.fillRect(0, 0, settings.canvas.width, settings.canvas.height);
    } else {
      this.ctx.clearRect(0, 0, settings.canvas.width, settings.canvas.height);
    }
  },

  draw_circ: function(x, y, r) {
    this.ctx.beginPath();
    this.ctx.arc(x, y, r, 0, Math.PI * 2, true);
    this.ctx.stroke();
    this.ctx.closePath();
  },

  draw_dot: function(x, y, r, stroke) {
    r = r || 3.5;
    this.ctx.beginPath();
    this.ctx.arc(x, y, r, 0, Math.PI * 2, true);
    this.ctx.fill();
    if (stroke) { this.ctx.stroke(); }
    this.ctx.closePath();
  },

  draw_line: function(x0, y0, x1, y1) {
    this.ctx.beginPath();
    this.ctx.moveTo(x0, y0);
    this.ctx.lineTo(x1, y1);
    this.ctx.stroke();
    this.ctx.closePath();
  },

});


/* Plotter with relative origin */

var Plotter = function(ctx, x, y) {
  this.set_origin(x, y);
  Plotter.prototype.__proto__.constructor.apply(this, [ctx]);
};
Plotter.prototype = Object.create(CanvasRenderer.prototype);
_.extend(Plotter.prototype, {
  constructor: Plotter,

  draw: function(theta, r) {
    this.draw_polar_point(theta, r);
  },

  refresh: function(theta, r) {
    this.reset();
    this.draw(theta, r);
  },

  set_origin: function(x, y) {
    this.origin = {x: x, y: y};
  },

  set_point_color: function(color) {
    this.point_color = color;
  },

  /* Convenience drawing methods */

  draw_polar_point: function(theta, radius) {
    this.draw_point(radius * Math.cos(theta), radius * Math.sin(theta));
  },

  draw_point: function(x, y) {
    this.ctx.strokeStyle = '#ccc';
    this.ctx.lineWidth = 1.5;
    this.draw_line_to_origin(x, y);
   
    this.ctx.fillStyle = this.point_color || settings.point.color;
    if (options.rave_mode) {  // head
      this.draw_glowing_dot(x, y, settings.point.size, this.ctx.fillStyle);
    } else {  // handle
      this.draw_dot(x, y, settings.point.size);
    }
  },

  draw_glowing_dot: function(x, y, r, color) {
    this.ctx.save();

    // blur spread doesn't work very well; let's draw a larger circle behind
    this.ctx.shadowBlur = settings.rave_mode.glow_blur;
    this.ctx.shadowOffsetX = 0;
    this.ctx.shadowOffsetY = 0;
    this.ctx.shadowColor = color;
    this.ctx.fillStyle = color;
    this.ctx.strokeStyle = color;
    this.draw_dot(x, y, r + settings.rave_mode.glow_spread, true);

    this.ctx.restore();
    this.ctx.fillStyle = '#fff';
    this.draw_dot(x, y, r);
  },

  draw_line_to_origin: function(x, y) {
    this.draw_line(0, 0, x, y);
  },

  /* Overridden primitives */

  draw_circ: function(x, y, r) {
    Plotter.prototype.__proto__.draw_circ.apply(this, [
      x + this.origin.x, y + this.origin.y, r,
    ]);
  },

  draw_dot: function(x, y, r, stroke) {
    Plotter.prototype.__proto__.draw_dot.apply(this, [
      x + this.origin.x, y + this.origin.y, r, stroke
    ]);
  },

  draw_line: function(x0, y0, x1, y1) {
    Plotter.prototype.__proto__.draw_line.apply(this, [
      x0 + this.origin.x, y0 + this.origin.y,
      x1 + this.origin.x, y1 + this.origin.y,
    ]);
  },
  
});


/* Polar plotter whose origin travels according to a periodic function.
 *  Traveling function returns a 2D Vector which offsets from
 *  initial origin.
 */

var TravelingPlotter = function(ctx, x, y, patterns) {
  this.patterns = patterns;  // this shouldn't change size
  this.initial_origin = {x: x, y: y};
  TravelingPlotter.prototype.__proto__.constructor.apply(this, [ctx, x, y]);
};
TravelingPlotter.prototype = Object.create(Plotter.prototype);
_.extend(TravelingPlotter.prototype, {
  constructor: TravelingPlotter,

  draw: function(theta, r) {
    this.reset();

    var self = this,
        args = arguments;
    _.each(this.get_patterns(), function(pattern, i) {
      var my_theta = theta + pattern.pattern_phase_shift;
      var color = settings.point_colors[i % settings.point_colors.length];
      self.set_traveling_origin(pattern.traveling_function, my_theta, r);
      self.set_point_color(color);

      if (window.options.show_hand_trace) {
        self.trace_origin(pattern, r);
      }
      if (window.options.show_pattern_trace) {
        self.trace_pattern(pattern, r, color);
      }

      // DEBUG: near relative theta=0, flash a different color
      // approximate because of step rounding errors caused by speed change
      /*if (Math.abs(pattern.shift_theta(my_theta)) <= get_d_theta()) {
      //  self.set_point_color(settings.POINT_DEBUG_COLOR);
      }*/

      // Draw the point
      self.constructor.prototype.__proto__.draw.apply(self, [
        pattern.shift_theta(my_theta), r * options.poi_ratio,
      ]);

      // todo: maybe make sure origins are drawn on top since they're smaller?
      self.draw_origin();
    });
  },

  // get non-null patterns
  get_patterns: function() {
    return _.filter(this.patterns, function(p) { return p; });
  },

  draw_origin: function() {
    this.ctx.fillStyle = settings.origin.color;
    this.draw_dot(0, 0, settings.origin.size);
  },

  // TODO: cache this since it only changes when the patterns do
  // TODO: light trails that decompose
  trace_pattern: function(pattern, r, color) {
    // Persistent trace of a pattern
    this.ctx.strokeStyle = color || settings.point.stroke_color;
    this.ctx.globalAlpha = (!options.rave_mode
                            ? settings.point.stroke_transparency
                            : settings.rave_mode.stroke_transparency
                           );

    var circle_fn = function_generators.circle();
    this.trace_function(
      function(theta, r) {
        return pattern.traveling_function(theta, r).add(
               circle_fn(pattern.shift_theta(theta), r * options.poi_ratio));
      }, r);

    this.ctx.globalAlpha = 1;
  },

  trace_origin: function(pattern, r, color) {
    // Persistent trace of the origin's movement
    this.ctx.strokeStyle = color || settings.origin.color;
    this.ctx.globalAlpha = settings.origin.stroke_transparency;

    this.trace_function(pattern.traveling_function, r);

    this.ctx.globalAlpha = 1;
  },

  trace_function: function(fn, r) {
    // Persistent trace of a function(theta, r)
    this.ctx.beginPath();
    this.ctx.lineWidth = 1.5;
    
    // simulate a full period
    // one extra point for smoothness
    for (var theta = 0; theta <= 2 * Math.PI; theta += get_d_theta()) {
      var coords = fn(theta, r);
      if (theta==0) {
        this.ctx.moveTo(this.initial_origin.x + coords.x,
                        this.initial_origin.y + coords.y);
      } else {
        this.ctx.lineTo(this.initial_origin.x + coords.x,
                        this.initial_origin.y + coords.y);
      }
    }
    this.ctx.stroke();
    this.ctx.closePath();

    // debug - highlight point @ theta=0
    if (options.DEV) {
      this.ctx.save();
      this.ctx.fillStyle = settings.POINT_DEBUG_COLOR;
      var coords = fn(0,r);
      Plotter.prototype.__proto__.draw_dot.apply(this, [
        this.initial_origin.x + coords.x, this.initial_origin.y + coords.y, 3, true]);
      this.ctx.restore();
    }
  },

  set_traveling_origin: function(traveling_function, theta, r) {
    var travel = traveling_function(theta, r),
        origin = [this.initial_origin.x + travel.x,
                  this.initial_origin.y + travel.y];
    this.set_origin.apply(this, origin);
  },

});


/* Traveling function signature:
 * :param theta: angle in radians
 * :param d: poi length (handle center to head center)
 * :returns: 2D Vector
 */
var traveling_functions = {

  // simple back and forth oscillation
  test: function(theta, d) {
    return new Vector(theta < Math.PI ? theta : 2*Math.PI - theta, 0);
  },

  // Circle phase shifted by π
  isolation: function(theta, r) {
    theta += Math.PI;
    return new Vector(r * Math.cos(theta), r * Math.sin(theta));
  },

};


// Parametric generators that return functions fitting the format above.
var function_generators = {

  /* :param n: number of sides
   * :param rotation: angle to phase shift, in radians, from pointing up
   * :param phase_shift: Math.PI/n to flip an antispin to an inspin
   */
  polygon: function(n, phase_shift, rotation) {
    phase_shift = phase_shift || 0
    rotation = rotation || 0;
    //if (!is_phase_shifted) rotation += Math.PI/n;
    // oh god why won't these rotate right

    return function(theta, d) {
      theta += rotation;
      // polar equation for edge: cos(π/n) / cos((theta mod 2π/n) - π/n)
      var r = (d * (
        Math.cos(Math.PI/n) / Math.cos((theta + phase_shift) % (2*Math.PI/n) - Math.PI/n)
      ));
      return new Vector(r * Math.cos(theta), r * Math.sin(theta));
    };
  },
  
  circle: function(frequency) {
    frequency = frequency || 1;
    return function(theta, r) {
      theta *= frequency;
      return new Vector(r * Math.cos(theta), r * Math.sin(theta));
    };
  },

};


// Complete pattern specifications
var patterns = {

  test: {
    frequency: -2,
    traveling_function: traveling_functions.isolation,
  },

  extension: {
    frequency: 1,
    traveling_function: function_generators.circle(),
  },

};


// struct for pattern specifications
function Pattern(options) {
  var self = this;
  // defines hand movement. required arg
  this.traveling_function = options.traveling_function;
  // poi frequency
  this.frequency = options.frequency || 1;
  // phase shift of poi from hand
  this.phase_shift = options.phase_shift || 0;
  // rotation of poi
  this.rotation = options.rotation || 0;
  // phase shift of entire pattern
  this.pattern_phase_shift = options.pattern_phase_shift || 0;
  // beats per period
  this.beats = Math.abs(this.frequency);

  this.shift_theta = function(theta, pattern) {
    /* This one is used to phase shift the poi */
    return (
      ((theta + self.phase_shift) * self.frequency + self.rotation) % (2*Math.PI));
  };
}


// Pattern generators
var pattern_generators = {

  n_petal_antispin: {
    generator: function(n, rotation, phase) {
      if (!rotation) { rotation = 0; }
      rotation = to_radians(rotation) + settings.UNIVERSAL_ROTATION;
      phase = to_radians(phase);
      var half_side_rotation = Math.PI / (2*n);
      return new Pattern({
        frequency: -(n-1),
        pattern_phase_shift: -Math.PI/(2*n) + phase,  // compensation for traveling offset
        phase_shift: half_side_rotation,
        rotation: -rotation * n,
        traveling_function: function_generators.polygon(n, rotation, half_side_rotation),
      });
    },
    args: [
      { name: 'N',
        default: 4,
        type: 'int',
        start: 3,
        stop: 8,
      },
      // todo: instead of absolute rotation, maybe specify amount out of phase
      // (fraction of π/n)
      { name: 'rotation',
        default: 0,
        type: 'int',
        values: [0, 30, 45, 60],
      },
      // pattern start point
      { name: 'phase',
        default: 0,
        type: 'int',
        start: 0,
        stop: 270,
        step: 90,
      },
    ],
  },

  n_petal_inspin: {
    // Note that the traveling function of an inspin is offset Math.PI/2n from
    // an antispin
    generator: function(n, rotation, phase) {
      rotation = to_radians(rotation) + settings.UNIVERSAL_ROTATION;
      phase = to_radians(phase);
      return new Pattern({
        frequency: -(n-1),
        rotation: -rotation * n,
        pattern_phase_shift: phase,
        traveling_function: function_generators.polygon(n, Math.PI/n + rotation, 0),
      });
    },
    args: [
      { name: 'N',
        default: 4,
        type: 'int',
        start: 3,
        stop: 8,
      },
      { name: 'rotation',
        default: 45,  // default to diamond mode
        type: 'int',
        values: [0, 30, 45, 60],
      },
      // pattern start point
      { name: 'phase',
        default: 0,
        type: 'int',
        start: 0,
        stop: 270,
        step: 90,
      },
    ],
  }

};


// struct for pattern generator: contains instructions for how to call it
function PatternGenerator(specs) {
  _.extend(this, specs);
}


// map specs to structs
patterns = _.chain(patterns).map(function(pattern, name) {
  return [name, new Pattern(pattern)];
}).object().value();
pattern_generators = _.chain(pattern_generators).map(function(gen, name) {
  return [name, new PatternGenerator(gen)];
}).object().value();
